Escape combining characters in char::Debug

varkor opened the pull request:
Although combining characters are technically printable, they make little sense to print on their own with `Debug`: it'd be better to escape them like non-printable characters.

This is a breaking change, but I imagine the fact `escape_debug` is rare and almost certainly primarily used for debugging that this is an acceptable change.
Resolves #41922.

r? @alexcrichton 
cc @clarcharr

clarcharr commented:
LGTM although the character ranges should probably be generated with a script instead of hard-coded, as they can change across unicode versions.

varkor commented:
@clarcharr: have you got any ideas for how to detect ranges programmatically? Apart from parsing the Unicode symbol names (which does not seem robust), I couldn't find any reasonable suggestion for how to do this automatically.

SimonSapin commented:
This PR is modifying the `char::escape_debug`. When calling this method directly or through `impl Debug for char`, there is only one code point being printed so the idea of escaping code points that do not make much sense on their own seems fine.

However this method is also used in `impl Debug for str` and in `str::escape_debug`. #41922 itself does not expect the behavior to change there and I agree. However that means making them based on something other than `char::escape_debug`. Should that other thing be exposed in new public API?

----

Regarding this being a breaking change, I think I’d be ok with documenting that the exact set of code points being escaped or not is not stable outside of the ASCII range.

The current doc-comment is already out of sync with the implementation: it claims that anything non-ASCII is escaped, which hasn’t been the case since https://github.com/rust-lang/rust/pull/34485.

----

I also agree with not hard-coding code point ranges, and instead extracting more Unicode data in `src/libstd_unicode/unicode.py`. However it’s unclear to me what exactly should be included. `Canonical_Combining_Class` is about reordering during normalization and probably not what we want. The [`Diacritic`](https://unicode.org/reports/tr44/#Diacritic) property is defined as:

> Characters that linguistically modify the meaning of another character to which they apply. Some diacritics are not combining characters, and some combining characters are not diacritics.

Then there’s the [`*Mark`](https://unicode.org/reports/tr44/#GC_Values_Table) properties:

> | Abbr | Long | Description |
> | -- | -- | -- |
> | Mn | Nonspacing_Mark | a nonspacing combining mark (zero advance width) |
> | Mc | Spacing_Mark | a spacing combining mark (positive advance width) |
> | Me | Enclosing_Mark | an enclosing combining mark |
> | M | Mark | Mn \| Mc \| Me |

However #41922 suggests that only nonspacing marks are problematic?

clarcharr commented:
I'd be fine with just going for nonspacing marks.

As for `str::escape_debug` goes, I'd personally opt for escaping any combination of nonspacing marks at the beginning of the string, although it's unclear if that would actually solve the problem. I'm fine with just leaving it with the current behaviour for now.

varkor commented:
> However that means making them based on something other than `char::escape_debug`. Should that other thing be exposed in new public API?

Yeah, I'm not sure about this. It'd be nicer if there was an extra argument (`single: bool`), so the behaviour could be different for individual characters, but we obviously can't change the signature. A specific public extra method that _just_ special-cases non-spacing combining characters seems a little specialised, though.

> The current doc-comment is already out of sync with the implementation: it claims that anything non-ASCII is escaped, which hasn’t been the case since #34485.

Which doc comment are you referring to here?

> I also agree with not hard-coding code point ranges, and instead extracting more Unicode data in `src/libstd_unicode/unicode.py`

I have a change that uses `Mn` instead of hard-coding ranges, pending the decision about the method / effect on `str` impls.

varkor commented:
@SimonSapin: as far as I can tell, there's an issue using `src/libstd_unicode/unicode.py` here, as it's not available for `core`, which is where the formatting trait is defined. Is it worth moving it, or taking a different approach?

clarcharr commented:
Wasn't the goal of having `libstd_unicode` separate from `libcore` to avoid having to include Unicode tables in libcore? Which makes it suddenly interesting considering how the changes to `escape_debug` make that less necessary. Or, I'm probably missing something.

Personally I really dislike the `libstd_unicode` façade, as it tends to get way less attention than libcore, liballoc, and libstd, which leads to bugs like #44303. I'm still confused why it exists and I don't think I've actually seen it explained somewhere.

SimonSapin commented:
https://github.com/rust-lang/rust/pull/49698 merges `std_unicode` into `core`.

It might be interesting to replace `printable.py` (formerly `char_private.py`) and `unicode.py` with something based on https://github.com/BurntSushi/ucd-generate.

bors commented:
:umbrella: The latest upstream changes (presumably #49698) made this pull request unmergeable. Please resolve the merge conflicts.

varkor commented:
@SimonSapin: #49698 makes everything much more pleasant :)

I've rebased on top of it (and also cleaned up `unicode.py` whilst I was at it, because it had some slightly annoying behaviour before).

SimonSapin commented:
@varkor Thanks for your work on this. However I’m still not sure what the right thing to do here is, in terms of what characters exactly to escape or not in which context. Do you know what other languages do for debug-printing Unicode strings and characters?

varkor commented:
@SimonSapin: [Swift does something similar](https://github.com/rust-lang/rust/issues/41922#issuecomment-375290541) to the proposed change here. I haven't tried working out exactly which characters they choose to escape, but it seems reasonable to assume they choose either the same, or a similar category.

varkor commented:
Ah, so Swift actually escapes any non-ASCII character, which is what Rust used to do before https://github.com/rust-lang/rust/issues/24588:
https://github.com/apple/swift/blob/c35d508600516b892732e2fd3f0f0a17ca4562ba/stdlib/public/core/UnicodeScalar.swift#L197-L252
So this is new territory, and it makes sense to go with what we think is most sensible: in this case, I think the suggestion of `Mn` is a reasonable choice.

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365822829) ([raw log](https://api.travis-ci.org/v3/job/365822829/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (615222/615222), completed with 4953 local objects.
---
[00:00:46] configure: rust.quiet-tests     := True
---
[00:42:10] error: character literal may only contain one codepoint
[00:42:10]  --> <stdin>:2:9
[00:42:10]   |
[00:42:10] 2 | assert!('\u{301}'.is_control());
[00:42:10]   |         ^^^^^^^^^
[00:42:10] help: if you meant to write a `str` literal, use double quotes
[00:42:10]   |
[00:42:10] 2 | assert!("\u{301}".is_control());
---
[00:44:13] ...............................................................................i....................
[00:44:19] ......................i.............................................................................
---
[00:45:01] ..i...........................................................................i.....................
---
[00:45:57] .............................................i......................................................
---
[00:49:53] .............................i......................................................................
[00:50:07] ..............................................................i.....................................
[00:50:22] ................................................i...................................................
[00:50:43] ....................................................................................................
[00:51:04] ....................................................................................................
[00:51:26] ....................................................................................................
[00:51:51] ......i.............................................................................................
[00:52:19] ..i.................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:52:27] ................
[00:52:59] ....................................................................................................
[00:53:33] ....................................................................ii..............................
[00:54:21] ...............................i....................................................i.ii..test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:54:26] ..........
[00:55:07] ............................................................................................iiiiiii.
---
[00:57:22] ...................i............................................................ii.iii..............
[00:57:29] ....................................................................................................
[00:57:36] .........i..............................i...........................................................
[00:57:44] ....................................................................................................
[00:57:50] .....................i..............................................................................
[00:57:58] ....................................................................................................
[00:58:08] ....................................................................................................
[00:58:17] ....................................................................................................
[00:58:28] ....................................................................................................
[00:58:41] ....................................................................................................
[00:58:50] ..............i.....................................................................................
[00:58:59] ..................i..ii.............................................................................
[00:59:09] ....................................................................................................
[00:59:18] ....................................................................................................
[00:59:27] .....................................................................................i..............
[00:59:37] ...............................i....................................................................
---
[01:00:12] ...........................i........................................................................
[01:00:13] ....................................................................i...............................
[01:00:15] ................i.......................................................
---
[01:00:29] .............i........................
---
[01:00:59] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:01:02] i.......i......................i.......
---
[01:01:39] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:01:40] ....ii...
---
[01:11:05] ......i.............................................................................................
---
[01:13:00] .....................................i..............................................................
[01:13:19] ....................................................................................................
[01:13:39] ..........................................i.........................................................
---
[01:15:03] ....................................................................F...............................
[01:15:03] ....................................................................................................
[01:15:03] ....................................................................................................
[01:15:03] ..........................................................ii........................................
---
[01:15:04] ---- char::test_debug stdout ----
[01:15:04]  thread 'char::test_debug' panicked at 'assertion failed: `(left == right)`
[01:15:04]   left: `"́"`,
[01:15:04]  right: `"\\u{301}"`', libcore/../libcore/tests/char.rs:193:5
---
[01:15:04] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
[01:15:04] expected success, got: exit code: 101
[01:15:04]
[01:15:04]
[01:15:04] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:15:04] Build completed unsuccessfully in 0:32:05
[01:15:04] make: *** [check] Error 1
[01:15:04] Makefile:58: recipe for target 'check' failed
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:067bf24a:start=1523572673261371199,finish=1523572673267437982,duration=6066783
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:0faba351
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:0faba351:start=1523572673273139769,finish=1523572673278684564,duration=5544795
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:069e8068
$ dmesg | grep -i kill
[   10.699412] init: failsafe main process (1095) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366222431) ([raw log](https://api.travis-ci.org/v3/job/366222431/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (614844/614844), completed with 4887 local objects.
---
[00:00:46] configure: rust.quiet-tests     := True
---
[00:45:16] ...............................................................................i....................
[00:45:22] ......................i.............................................................................
---
[00:46:06] ..................i...........................................................................i.....
[00:46:13] ....................................................................................................
[00:46:20] ........ii..........................................................................................
---
[00:47:08] .............................................i......................................................
---
[00:51:16] .............................i......................................................................
[00:51:33] ...............................................................i....................................
[00:51:50] .................................................i..................................................
[00:52:13] ....................................................................................................
[00:52:37] ....................................................................................................
[00:53:01] ....................................................................................................
[00:53:29] ......i.............................................................................................
[00:53:53] ..i.............................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:54:09] ....................................
[00:54:45] ....................................................................................................
[00:55:21] ....................................................................ii..............................
[00:56:10] ...............................i.........................................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:56:21] ...........i.ii............
[00:57:07] ............................................................................................iiiiiii.
---
[00:59:40] ...................i............................................................ii.iii..............
[00:59:49] ....................................................................................................
[00:59:58] ........i..............................i............................................................
[01:00:07] ....................................................................................................
[01:00:14] ..........i.........................................................................................
[01:00:24] ....................................................................................................
[01:00:35] ....................................................................................................
[01:00:46] ....................................................................................................
[01:00:58] ....................................................................................................
[01:01:12] ....................................................................................................
[01:01:21] ...i................................................................................................
[01:01:31] .......i..ii........................................................................................
[01:01:43] ....................................................................................................
[01:01:54] ....................................................................................................
[01:02:04] .........................................................................i..........................
[01:02:16] ...................i................................................................................
---
[01:02:55] ...........................i........................................................................
[01:02:57] ....................................................................i...............................
[01:02:59] ................i.......................................................
---
[01:03:17] .............i........................
---
[01:03:54] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:03:58] i.......i......................i.......
---
[01:04:47] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:04:49] ....ii...
---
[01:14:37] ......i.............................................................................................
---
[01:16:49] .....................................i..............................................................
[01:17:10] ....................................................................................................
[01:17:34] ..........................................i.........................................................
---
[01:19:16] ..................................................................F.................................
[01:19:16] ....................................................................................................
[01:19:16] ....................................................................................................
[01:19:16] ..........................................................ii........................................
---
[01:19:16] ---- char::test_debug stdout ----
[01:19:16]  thread 'char::test_debug' panicked at 'assertion failed: `(left == right)`
[01:19:16]   left: `"\'a\'"`,
[01:19:16]  right: `"a"`', libcore/../libcore/tests/char.rs:191:5
---
[01:19:16] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
[01:19:16] expected success, got: exit code: 101
[01:19:16]
[01:19:16]
[01:19:16] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:19:16] Build completed unsuccessfully in 0:35:15
[01:19:16] make: *** [check] Error 1
[01:19:16] Makefile:58: recipe for target 'check' failed
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:04448238:start=1523645980005461977,finish=1523645980016987409,duration=11525432
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:1593f37b
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:1593f37b:start=1523645980026340490,finish=1523645980040414857,duration=14074367
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0dccad20
$ dmesg | grep -i kill
[   11.720930] init: failsafe main process (1097) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366288294) ([raw log](https://api.travis-ci.org/v3/job/366288294/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (614848/614848), completed with 4887 local objects.
---
[00:00:53] configure: rust.quiet-tests     := True
---
[00:58:55] ...............................................................................i....................
[00:59:03] ......................i.............................................................................
---
[00:59:54] ...................i...........................................................................i....
[01:00:01] ....................................................................................................
[01:00:10] .........ii.........................................................................................
---
[01:01:10] .............................................i......................................................
---
[01:06:19] .............................i......................................................................
[01:06:36] ..............................................................i.....................................
[01:06:55] ................................................i...................................................
[01:07:20] ....................................................................................................
[01:07:47] ....................................................................................................
[01:08:15] ....................................................................................................
[01:08:46] ......i.............................................................................................
[01:09:06] ..i..........................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[01:09:33] .......................................................
[01:10:13] ....................................................................................................
[01:10:56] ....................................................................ii..............................
[01:11:41] ...............................i.......................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[01:12:04] .............................i.ii............
[01:12:55] ............................................................................................iiiiiii.
---
[01:16:30] ...................i............................................................ii.iii..............
[01:16:39] ....................................................................................................
[01:16:49] ........i..............................i............................................................
[01:16:58] ....................................................................................................
[01:17:06] ..........i.........................................................................................
[01:17:16] ....................................................................................................
[01:17:28] ....................................................................................................
[01:17:40] ....................................................................................................
[01:17:52] ....................................................................................................
[01:18:08] ....................................................................................................
[01:18:09] ...i.test [compile-fail] compile-fail/issue-22638.rs has been running for over 60 seconds
[01:18:20] ...............................................................................................
[01:18:32] .......i..ii........................................................................................
[01:18:44] ....................................................................................................
[01:18:55] ....................................................................................................
[01:19:05] .........................................................................i..........................
[01:19:18] ...................i................................................................................
---
[01:19:56] ............................i.......................................................................
[01:19:58] ....................................................................i...............................
[01:19:59] ................i.......................................................
---
[01:20:18] .............i........................
---
[01:20:54] i....i.ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:20:58] i.......i......................i.......
---
[01:21:49] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:21:50] ....ii...
---
[01:34:43] ......i.............................................................................................
---
[01:37:24] .....................................i..............................................................
[01:37:54] ....................................................................................................
[01:38:25] ..........................................i.........................................................
---
[01:40:27] ..........................................................ii........................................
---
[01:41:04] ...............................................................................................F....
[01:41:25] ....................................................................................................
[01:41:54] ....................................................................................................
[01:42:19] ......................i.............................................................................
---
[01:46:45] ---- char/methods.rs - char::methods::char::is_nonspacing_mark (line 704) stdout ----
[01:46:45]  error[E0658]: use of unstable library feature 'rustc_private': mainly needed for compiler internals (see issue #27812)
[01:46:45]  --> char/methods.rs:706:19
[01:46:45]   |
[01:46:45] 5 | assert!('\u{301}'.is_nonspacing_mark());
[01:46:45]   |                   ^^^^^^^^^^^^^^^^^^
[01:46:45]   |
[01:46:45]   = help: add #![feature(rustc_private)] to the crate attributes to enable
[01:46:45]
[01:46:45] error[E0658]: use of unstable library feature 'rustc_private': mainly needed for compiler internals (see issue #27812)
[01:46:45]  --> char/methods.rs:707:14
[01:46:45]   |
[01:46:45] 6 | assert!(!'e'.is_nonspacing_mark());
[01:46:45]   |              ^^^^^^^^^^^^^^^^^^
[01:46:45]   |
[01:46:45]   = help: add #![feature(rustc_private)] to the crate attributes to enable
[01:46:45]
[01:46:45] thread 'char/methods.rs - char::methods::char::is_nonspacing_mark (line 704)' panicked at 'couldn't compile the test', librustdoc/test.rs:313:13
---
[01:46:45] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
[01:46:45] expected success, got: exit code: 101
[01:46:45]
[01:46:45]
[01:46:45] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:46:45] Build completed unsuccessfully in 0:49:27
[01:46:45] make: *** [check] Error 1
[01:46:45] Makefile:58: recipe for target 'check' failed
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:10d684e6:start=1523657720414955872,finish=1523657720423984553,duration=9028681
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:0564d390
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:0564d390:start=1523657720430898101,finish=1523657720437594237,duration=6696136
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0b400793
$ dmesg | grep -i kill
[   10.816061] init: failsafe main process (1095) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

TimNN commented:
@SimonSapin: Friendly triage ping :)

pietroalbini commented:
Ping from triage! Can @SimonSapin or someone else from @rust-lang/libs review this?

pietroalbini commented:
Ping from triage! This PR still needs a review! Can @SimonSapin or someone else from @rust-lang/libs review this?

shepmaster commented:
Ping from triage, @aturon. This PR has not been reviewed by @rust-lang/libs in multiple weeks.

SimonSapin commented:
More than code review, this needs a design decision

alexcrichton commented:
@SimonSapin that's a decision for @rust-lang/libs, right? When you get a chance, could you (or @varkor) summarize the decision that needs to be made?

alexcrichton commented:
Ok thanks! I'll make sure this comes up at the next libs triage

SimonSapin commented:
In my mind the question is rather: what is the exact set of code points that should be escaped or not? Unicode has a number of "properties" (see above) that seem relevant, but it’s not clear to me which one (or which combination) is most appropriate.

SimonSapin commented:
Also, `char::escape_debug` is used both in:

* `impl Debug for char` where we only ever have a single code point, so a combining code point never has anything to combine with. So this change makes sense
* `impl Debug for str` (for each `char` in the string) where we probably do *not* want that same behavior. Maybe?

But then if these two impls want to escape different sets of code points, should there be two methods like `char::escape_debug`?

varkor commented:
@SimonSapin: note that I changed the implementation so that only the `Debug` impl for `char` is affected — `Debug` for `str` is entirely unchanged. So now this is purely a formatting change. It's unclear what `char::escape_debug` is used for in general though, so we might want to change it too.

Manishearth commented:
An alternative to escaping them is prepending them with a zero width non breaking space or just a regular non breaking space, which forces them to render as their uncombined form.

clarcharr commented:
@Manishearth in the original issue I recommended printing a dotted circle (U+25CC) which is generally used for demonstrating combining marks, although it was preferred to just escape them so that the result remains valid Rust.

A zero-width space might help for debug-printing strings, though.

Manishearth commented:
The dotted circle version of combining chars is what they do in isolation
when they start text (or after any kind of non breaking space), I don't
think it's a good idea to combine them with the dotted circle code point to
get this. That code point won't be rescaled based on the size of the
combiner, it's not very good for rendering combiners. Use non breaking
spaces if you want a dotted circle to show.

(Some fonts don't use dotted circles for combiners but they render
something similarly sensible)

On Mon, May 14, 2018, 4:56 PM Clar Roʒe <notifications@github.com> wrote:

> @Manishearth <https://github.com/Manishearth> in the original issue I
> recommended printing a dotted circle (U+25CC) which is generally used for
> demonstrating combining marks, although it was preferred to just escape
> them so that the result remains valid Rust.
>
> A zero-width space might help for debug-printing strings, though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/49283#issuecomment-388975747>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABivSC-hy8RfbQYTvL-98yQGln1ghGRJks5tyf2BgaJpZM4S3jbN>
> .
>

SimonSapin commented:
The libs team discussed this today, consensus was that we’d prefer to keep `Debug for char`, `Debug for str`, and `escape_debug` all consistent with each other. The way to do that while fixing #41922 is to treat `char` as a one code point string, and escape combining marks when they’re the *first code point* in a string.

@Manishearth mentioned that the `Grapheme_Extend` derived property might be more appropriate to determine which extra code points to escape.

The PR currently adds a public unstable method `char::is_nonspacing_mark`, but I’d prefer to have its users call the crate-private function directly instead.

@varkor Could you make these changes?

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/379949107) ([raw log](https://api.travis-ci.org/v3/job/379949107/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://files.pythonhosted.org/packages/14/0e/00d0126afe94c8ea1dacf925d96d9cfcc0eb8f6a910689a9acfbee99f72b/awscli-1.15.22-py2.py3-none-any.whl (1.3MB)
    0% |▎                               | 10kB 23.0MB/s eta 0:00:01
    1% |▌                               | 20kB 2.5MB/s eta 0:00:01
    2% |▉                               | 30kB 2.7MB/s eta 0:00:01
    3% |█                               | 40kB 2.3MB/s eta 0:00:01
---
[00:59:59] .........test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[01:00:30] ...........................................................................................
[01:00:52] .......................................................................................ii...........
[01:01:48] ...................................................i................................................
[01:01:55] ....i.ii...........test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[01:02:50] ............iiiiiii.................................................................................
[01:03:12] ....................................................................................................
[01:03:33] ....................................................................................................
[01:03:56] .....................................................................................
---
[01:18:02] ....................................................................................................
[01:18:02] thread '<unnamed>' panicked at 'explicit panic', liballoc/../liballoc/tests/slice.rs:1306:17
[01:18:02] ....................................................................................................
[01:18:02] ....................................................................................................
[01:18:02] .......F............................................................................................
[01:18:05] ................................
[01:18:05] failures:
[01:18:05] 
[01:18:05] ---- str::test_escape_debug stdout ----
[01:18:05] ---- str::test_escape_debug stdout ----
[01:18:05]  thread 'str::test_escape_debug' panicked at 'assertion failed: `(left == right)`
[01:18:05]   left: `"\\u{301}a\\u{301}bé\\u{e000}"`,
[01:18:05]  right: `"\u{301}a\u{301}bé\u{e000}"`', liballoc/../liballoc/tests/str.rs:1002:5
[01:18:05] 
[01:18:05] failures:
[01:18:05]     str::test_escape_debug
[01:18:05] 
[01:18:05] 
[01:18:05] test result: FAILED. 531 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:18:05] 
[01:18:05] error: test failed, to rerun pass '--test collectionstests'
[01:18:05] 
[01:18:05] 
[01:18:05] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "alloc" "--" "--quiet"
[01:18:05] 
[01:18:05] 
[01:18:05] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:18:05] Build completed unsuccessfully in 0:29:35
[01:18:05] Build completed unsuccessfully in 0:29:35
[01:18:05] make: *** [check] Error 1
[01:18:05] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:055bfa96
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

varkor commented:
@SimonSapin: it would be great to finish this off before https://github.com/rust-lang/rust/pull/50922 and friends get merged, to avoid refactoring all over again.

clarcharr commented:
@vakor I was actually planning to rebase over these changes some time tomorrow so this is going to have to be merged earlier anyway.

varkor commented:
@clarcharr: ah, that sounds great! :)

SimonSapin commented on src/liballoc/str.rs:
```diff
@@ -377,12 +377,15 @@ impl str {
                issue = "27791")]
     pub fn escape_debug(&self) -> String {
-        self.chars().flat_map(|c| c.escape_debug()).collect()
+        self.chars().enumerate().flat_map(|(i, c)| c.escape_debug_ext(i == 0)).collect()
```
Please write this as follow to avoid counting code points:

```rust
    let mut string = String::with_capacity(self.len());
    let mut chars = self.chars();
    if let Some(first) = chars.next() {
        string.extend(first.escape_debug_ext(true))
    }
    string.extend(chars.flat_map(|c| c.escape_debug_ext(false)));
    string
```

SimonSapin commented on src/liballoc/tests/str.rs:
```diff
@@ -999,6 +999,7 @@ fn test_escape_debug() {
     assert_eq!("\u{10000}\u{10ffff}".escape_debug(), "\u{10000}\\u{10ffff}");
     assert_eq!("ab\u{200b}".escape_debug(), "ab\\u{200b}");
     assert_eq!("\u{10d4ea}\r".escape_debug(), "\\u{10d4ea}\\r");
+    assert_eq!("\u{301}a\u{301}bé\u{e000}".escape_debug(), "\\u{301}a\u{301}bé\\u{e000}");
```
(I’ve had to read this a couple times. Maybe add a comment above to call out the difference between `\u{}` and `\\u{}` in expected results?)

SimonSapin commented on src/libcore/unicode/tables.rs:
```diff
@@ -93,7 +93,7 @@ pub mod general_category {
         ],
         r4: [
-            0, 1, 2, 3, 3, 3, 4, 3, 3, 3, 3, 3, 3, 5, 6, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+            0, 1, 2, 3, 3, 3, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
```
It’s very worrying that existing tables change without the Unicode version number changing. Is this a bug in `unicode.py`? Or do you get the same changes by re-running unmodified `unicode.py`?

varkor replied:
Turns out that the files I downloaded before were truncated. Rerunning the version of `unicode.py` in the PR fixed it. Whoops.

SimonSapin commented on src/libcore/unicode/unicode.py:
```diff
@@ -21,11 +21,14 @@
 #
 # Since this should not require frequent updates, we just store this
-# out-of-line and check the unicode.rs file into git.
+# out-of-line and check the unicode.py file into git.
```
I’m pretty sure that the original comment did mean `.rs`, even though the generated file has since been renamed a couple times (now `tables.rs`).

varkor replied:
Ah, I thought it could have been either. I've updated it to `tables.rs`.

SimonSapin commented:
@bors r+

bors commented:
:pushpin: Commit b653937 has been approved by `SimonSapin`

<!-- @bors r=SimonSapin b6539372e9d4dddb1c9f8c894f23bd4c3e8d9489 -->

bors commented:
:hourglass: Testing commit b6539372e9d4dddb1c9f8c894f23bd4c3e8d9489 with merge 65a16c000720f286c91747b91e6d9caa7b1bb63c...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7491), [status-travis](https://travis-ci.org/rust-lang/rust/builds/381916317?utm_source=github_status&utm_medium=notification)
Approved by: SimonSapin
Pushing 65a16c000720f286c91747b91e6d9caa7b1bb63c to master...
